src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs

```diff
@@ -6,213 +6,188 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
     using System;
     using System.Diagnostics;
     using System.Linq;
-    using System.Threading;
     using System.Threading.Tasks;
-
     using Microsoft.Azure.ServiceBus.Core;
     using Xunit;
 
-    public class SessionDiagnosticsTests : DiagnosticsTests
+    public class SessionDiagnosticsTests : DiagnosticsTests, IDisposable
     {
-        private IMessageSession messageSession;
-        private SessionClient sessionClient;
-        private MessageSender messageSender;
-        private QueueClient queueClient;
-        private bool disposed;
-
         [Fact]
         [LiveTest]
         [DisplayTestMethodName]
-        async Task AcceptSetAndGetStateGetFireEvents()
+        public async Task AcceptSetAndGetStateGetFireEvents()
         {
             await ServiceBusScope.UsingQueueAsync(partitioned: false, sessionEnabled: true, async queueName =>
             {
-                this.messageSender = new MessageSender(TestUtility.NamespaceConnectionString, queueName);
-                this.sessionClient = new SessionClient(TestUtility.NamespaceConnectionString, queueName, ReceiveMode.ReceiveAndDelete);
-
-                this.listener.Enable();
-
-                var sessionId = Guid.NewGuid().ToString();
-                await this.messageSender.SendAsync(new Message
+                var messageSender = new MessageSender(TestUtility.NamespaceConnectionString, queueName);
+                var sessionClient = new SessionClient(TestUtility.NamespaceConnectionString, queueName, ReceiveMode.ReceiveAndDelete);
+                var messageSession = default(IMessageSession);
+                var eventQueue = this.CreateEventQueue();
+                
+                try
                 {
-                    MessageId = "messageId",
-                    SessionId = sessionId
-                });
-                this.messageSession = await this.sessionClient.AcceptMessageSessionAsync(sessionId);
+                    using (var listener = this.CreateEventListener(queueName, eventQueue))
+                    using (var subscription = this.SubscribeToEvents(listener))
+                    {
+                        listener.Enable();
```

Comment:
Seems excessive to have `CreateEventListener` and `SubscribeToEvents` and `listener.Enable` why do we need three of them instead of just one?

Reply:
Also CreateEventQueue. Can we return FakeDiagnosticListener instance that would automatically be subscribed to events and have an event queue as a property?

Reply:
This largely kept the same pattern from the existing code.  Collapsing the event queue into the listener is a good idea.  I'm on the fence with automatically subscribing.  The `Enable` method is poorly named (but prevalent in the existing code semantics) - it actually changes the filtering predicate for whether the listener is passed an activity or not.    By default, it filters everything out.

I'm inclined to leave this as the tests have very specific assumptions around what activities are present and their order.   While it is theoretically safe to refactor this and apply a different pattern, I'm not seeing it as a high value gain for the potential risk and effort.

Reply:
After looking over the structure, I'm inclined to leave this as-is.   The responsibility of the listener is determining if an activity is interesting to its consumers and, if so, invoking the delegate that takes action on the observation.   It does not have an understanding of what the consumer of that listener is doing.   

Were I writing this from scratch, I'd likely have gone with a more specialized listener that had more opinionated behavior and could collapse things down close to what you're suggesting.  In this case, however, I don't feel that there is enough benefit to refactoring to warrant the time and risk.